### ISSUE TITLE
Make the library compatible with client/server environments where window is undefined

### ISSUE THREAD

**Issue by mtt87** [aermin/web-locks#32]
👋  Thanks for building this project.

I propose the following change to make it compatible with client/server frameworks like next.js.
At the moment trying to use the library with next.js throws the error `window is not defined` because a page can be rendered on the server side where `window` is `undefined`.

I tested locally and it should work with 2 simple changes
### `index.ts`
```ts
const locks = (function () {
  // Add this block
  if (typeof window === "undefined") {
    return {
      request: async () => {},
      query: () => ({
        held: [],
        pending: [],
      }),
    } as Partial<LockManager>;
  }
  // existing code
})();
```

### `polyfill.ts`
```ts
  constructor() {
    this._defaultOptions = {
      mode: LOCK_MODE.EXCLUSIVE,
      ifAvailable: false,
      steal: false,
    };
    // Wrap init inside a check
    if (typeof window !== "undefined") {
      this._init();
    }
  }
```

**Comment by aermin** [comment]
Hi @mtt87,  thank you for your attention and advice, I refactor this polyfill usage, pr => https://github.com/aermin/web-locks/pull/33

I think if the environment did not exist  `window.navigator`,  let this polyfill not run is more appropriate. Maybe you could handle it on your side like the below code.

```ts
import "navigator.locks";

navigator?.locks.request("my_resource", async (lock) => {
  // The lock has been acquired.
  await do_something();
  await do_something_else();
  // Now the lock will be released.
});
```

**Comment by mtt87** [comment]
Thanks for the quick reply.

I tried PR #33 but it still doesn't work with `next.js`

```js
// works fine and console logs "undefined"
console.log(typeof window) 

// throws error "window is not defined"
console.log(window) 

// throws error "window is not defined"
const navigator = window?.navigator as Navigator & { locks: LockManager }; 
```

Using `window?.navigator` is equivalent of `(window === null || window === undefined) ? undefined : window` but it still throws an error, you can reproduce it in node
```
$ node
Welcome to Node.js v16.4.0.
Type ".help" for more information.
> window?.navigator
Uncaught ReferenceError: window is not defined
> window === null
Uncaught ReferenceError: window is not defined
> window === undefined
Uncaught ReferenceError: window is not defined
> 
```

The only way to fix it is to check `typeof window !== "undefined"` that will not throw.

**Comment by aermin** [comment]
Got it, thx for your grate advice. I update it in the [pr](https://github.com/aermin/web-locks/pull/33/files)  @mtt87 pls review  ^ ^

**Comment by mtt87** [comment]
👍  now the code works but I notice you removed the export of the types 😄  I think they should still be there as it helps people do stuff in Typescript
```ts
import {
  Lock,
  LockInfo,
  LockManager,
  LockManagerSnapshot,
  LocksInfo,
} from "./polyfill";

(function () {
  if (typeof window !== "undefined") {
    const navigator = window.navigator as Navigator & { locks: LockManager };
    if (navigator && !navigator.locks) {
      const lockManager = new LockManager();
      Object.defineProperty(navigator, "locks", {
        value: lockManager,
      });
    }
  }
})();

export type { LockManager, LockInfo, LocksInfo, Lock, LockManagerSnapshot };
```

Then in my codebase I can do
```ts
import "navigator.locks";
import type { LockManager } from "navigator.locks";

(navigator as Navigator & { locks: LockManager }).locks.request(...)
```

After this I think everything looks great 👍 🥳

**Comment by aermin** [comment]
ok bro, good advice !!   thanks, I push a new commit to the pr. If now is ok now,  I will merge and make a new release. @mtt87 

thanks again ^ ^

**Comment by mtt87** [comment]
Fantastic thank you 😃

**Comment by aermin** [comment]
@mtt87 I released the new version(0.3,1)  🥂
